zilliqa/src/consensus.rs

```diff
@@ -345,39 +361,47 @@ impl Consensus {
             info!(%peer_id, "peer does not have sufficient stake");
             return Ok(None);
         }
-        if let Some(existing) = self.pending_peers.iter_mut().find(|v| v.peer_id == peer_id) {
-            existing.public_key = public_key;
+        if let Some(existing) = self.peers.iter_mut().find(|v| v.public_key == public_key) {
+            existing.peer_id = Some(peer_id);
             info!(%peer_id, "peer already exists");
-            return Ok(None);
+            return Ok(Some((
+                Some(peer_id),
+                ExternalMessage::CommitteeJoined(self.public_key()),
+            )));
         }
         info!(%peer_id, "adding peer to consensus");
 
-        self.pending_peers.push(Validator {
-            peer_id,
+        self.peers.push(Validator {
+            peer_id: Some(peer_id),
             public_key,
         });
 
-        if self.view.get_view() == 1 {
-            let Some(genesis) = self.get_block_by_view(0)? else {
-                // if we don't have genesis that means we only have its hash
-                // ergo we weren't, and can't be, part of the network at genesis and
-                // can't vote for it anyway
-                return Ok(None);
-            };
-            // If we're in the genesis committee, vote again.
-            if genesis
-                .committee
-                .iter()
-                .any(|v| v.peer_id == self.peer_id())
-            {
-                trace!("voting for genesis block");
-                let leader = self.leader(&genesis.committee, self.view.get_view());
-                let vote = self.vote_from_block(&genesis);
-                return Ok(Some((Some(leader.peer_id), ExternalMessage::Vote(vote))));
-            }
+        Ok(Some((
+            Some(peer_id),
+            ExternalMessage::CommitteeJoined(self.public_key()),
+        )))
+    }
+
+    // This is a callback method that is called by a 'joining' node
+    // It happens in reply to JoinCommittee message (triggered via CommitteeJoined)
+    // It works as follows: Peer wants to join, it broadcasts JoinCommittee. Each committee member
+    // that receives this message replies with its own peer_id and pub_key by sending CommiteeJoined
```

Comment:
Maybe I'm misunderstanding some of the solution, but couldn't we store all of this data in the contract and remove the need for the two P2P messages?

Reply:
I can agree with this approach - the peer_id is a networking address and storing it in the contract feels wrong.

On the other hand, on further thought, the peer_id is derived from the node's private key anyway, just as the pubkey is. So a node with a given pubkey is always going to have the same peer_id; and the only reason we need to store both is because there's no way to derive one from the other without knowing the privkey. (There isn't, right? If there was one it'd be very nice.) So maybe storing it in the contract is fine? It'd reduce the complexity of the node's state machine, too. @JamesHinshelwood thoughts?

Reply:
The peer ID is a node's identity on the network. A single peer ID can have multiple IP addresses which change over the course of time - `libp2p` handles this mapping (at least that's my understanding anyway).

And yeah - The peer ID is a deterministic function of the secret key, just like the public key.

IMO we should store it in the contract state. We will have the two identities problem wherever we store them, because there are two ways of identifying a node, depending on if you are at the network level (peer ID) or at the consensus level (BLS public key). If we still keep some of the data _outside_ of the contract (like in the current state of this PR), then we still need network-level messages to maintain that data. If we move everything inside the contract, that complexity disappears and nodes don't even need to be able to 'join' themselves. A node can simply keep synced with the chain, notice when its own identity turns up in the consensus committee (because the user externally submitted a transaction to add it with some stake) and then start participating.

Reply:
My only thought is that conceptually, network and blockchain identities might be different. Sure, right now they're derived from the same private key, but what if someone modified their node software such that the peer ID uses a different private key - and then ran multiple nodes that use the same on-chain priv/pub key identity (and can therefore do things like authorize each other's transactions, participate in consensus together, etc.), but have different peer IDs?

But, yeah, on further thought I think this is bikeshedding and also needless complexity at this point. Other things will probably break already if you do this, and we can probably just say that we don't support it instead, and link both identities together at the protocol level and store them in the contract, eliminating the network complexity.